File: Algorithm_4th_Edition/ShellSort.java

public class ShellSort {
    //核心代码---开始
    public static void sort(int[] arr) {
        int j;
        for (int gap = arr.length / 2; gap >  0; gap /= 2) {
            for (int i = gap; i < arr.length; i++) {
                int tmp = arr[i];
                for (j = i; j >= gap && tmp < arr[j - 1]; j -= gap) {
                    arr[j] = arr[j - gap];
                }
                arr[j] = tmp;
            }
        }
    }
    //核心代码---结束
    public static void main(String[] args) {

        int N = 2000;
        int[] arr = {5, 3, 2, 1, 4};
        ShellSort.sort(arr);
        for( int i = 0 ; i < arr.length ; i ++ ){
            System.out.print(arr[i]);
            System.out.print(' ');
        }
    }
}

File: Algorithm_4th_Edition/shell_sort.py
import random
def shell_sort(arr: list):
    N = len(arr)
    h = 1
    while h < (N/3):
        h = 3 * h + 1
    while h >= 1:
        # h-sort the array
        for i in range(h, N):
            # insert arr[i] among a[i-h]
            j = i
            while j >= h and arr[j] < arr[j-h]:
                arr[j], arr[j-h] = arr[j-h], arr[j]
                j -= h
        h = h//3
    
    return arr

test = [i for i in range(100)]
random.shuffle(test)
print(shell_sort(test))
